aten/src/ATen/native/cuda/DistributionTemplates.h

```diff
@@ -557,45 +540,16 @@ struct GeometricKernel {
 
 template<typename RNG>
 void exponential_kernel(TensorIterator& iter, double lambda_, RNG gen) {
-  // Note that HIP doesn't support std::nextafter in device code.
-  auto nextafter_1_0_float = std::nextafter(1.0f, 0.0f);
-  auto nextafter_1_0_double = std::nextafter(1.0, 0.0);
   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "exponential_cuda", [&] {
     using accscalar_t = at::acc_type<scalar_t, true>;
     auto lambda = static_cast<accscalar_t>(lambda_);
-    if (std::is_same<scalar_t, double>::value) {
-      // define lambda for exponential transformation
-      auto exponential_func = [lambda, nextafter_1_0_double] __device__ (accscalar_t rand) {
-        accscalar_t sample;
-        // curand_uniform has (0,1] bounds. log(1) is 0 and exponential excludes 0.
-        // Hence, squash the 1 to just below 1.
-        if(rand == static_cast<accscalar_t>(1.0)) {
-          sample = ::log(nextafter_1_0_double);
-        } else {
-          sample = ::log(rand);
-        }
-        return static_cast<scalar_t>(static_cast<accscalar_t>(-1.0) / lambda * sample);
-      };
-      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls/2>(iter,
-        gen,
-        [] __device__ (curandStatePhilox4_32_10_t* state) { return curand_uniform2_double(state); },
-        exponential_func);
-    } else {
-      // use __logf fast approximation for peak bandwidth
-      auto exponential_func = [lambda, nextafter_1_0_float] __device__ (accscalar_t rand) {
-        accscalar_t sample;
-        if(rand == static_cast<accscalar_t>(1.0)) {
-          sample = __logf(nextafter_1_0_float);
-        } else {
-          sample = __logf(rand);
-        }
-        return static_cast<scalar_t>(static_cast<accscalar_t>(-1.0) / lambda * sample);
-      };
-      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls>(iter,
-        gen,
-        [] __device__ (curandStatePhilox4_32_10_t* state) { return curand_uniform4(state); },
-        exponential_func);
-    }
+    // define lambda for exponential transformation
+    auto exponential_func = [lambda] __device__ (accscalar_t rand) {
+      // curand_uniform has (0,1] bounds. Reverse it to exclude 1 which produces `+inf` after applying exponential transformation
```

Comment:
This way of handling the 1 case is different from what @fritzo describes [here](https://github.com/pytorch/pytorch/pull/37984/files#r425330494). This would increase the odds of getting 0, since 1 is also mapped to 0.

An interesting question is how to test this. Thoughts?

Reply:
Re: testing, I do like the general aim of this PR to separate transform logic from sampling logic, since we can then test those transforms on synthetic values (maybe 0, finfo.tiny, 1-finfo.eps, 1). Taking this approach to its logical conclusion, each transform would need to input an unmodified rng return value, whereas as this PR currently stands, the transform logic is still split between the caller (e.g. `auto reverse_bound_rand = ...`) and the transforms. (Note this suggestion only tests for extreme values, and does not test the frequency properties @ngimel mentions).

Reply:
Yeah, testing deserves a more detailed discussion. One small point I'd like to make is testing on something like finfo.tiny does not really make sense because it is impossible to get finfo.tiny the way uniformly distributed numbers are generated now (the smallest number is either 2^-32 or 2^-24).